amaranth/vendor/gowin.py

```diff
@@ -0,0 +1,549 @@
+        return oscillators[0]
+
+
+    ### Common templates
+
+    _common_file_templates = {
+        "{{name}}.v": r"""
+            /* {{autogenerated}} */
+            {{emit_verilog()}}
+        """,
+        "{{name}}.debug.v": r"""
+            /* {{autogenerated}} */
+            {{emit_debug_verilog()}}
+        """,
+        "{{name}}.cst": r"""
+            // {{autogenerated}}
+            {% for port_name, pin_name, attrs in platform.iter_port_constraints_bits() -%}
+                IO_LOC "{{port_name}}" {{pin_name}};
+                {% for attr_name, attr_value in attrs.items() -%}
+                    IO_PORT "{{port_name}}" {{attr_name}}={{attr_value}};
+                {% endfor %}
+            {% endfor %}
+        """,
+        "{{name}}.sdc": r"""
+        // {{autogenerated}}
+        {% for net_signal,port_signal,frequency in platform.iter_clock_constraints() -%}
+            create_clock -name {{port_signal.name|tcl_escape}} -period {{1000000000/frequency}}  [get_ports {{port_signal.name|tcl_escape}}]
+        {% endfor %}
+        {{get_override("add_constraints")|default("# (add_constraints placeholder)")}}
+        """,
+    }
+
+
+    ### Apicula templates
+
+    _apicula_required_tools = ["yosys", "nextpnr-gowin", "gowin_pack"]
+
+    _apicula_file_templates = {
+        **TemplatedPlatform.build_script_templates,
+        **_common_file_templates,
+        "{{name}}.il": r"""
+            # {{autogenerated}}
+            {{emit_rtlil()}}
+        """,
+        "{{name}}.ys": r"""
+            # {{autogenerated}}
+            {% for file in platform.iter_files(".v") -%}
+                read_verilog {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
+            {% for file in platform.iter_files(".sv") -%}
+                read_verilog -sv {{get_override("read_verilog_opts")|options}} {{file}}
+            {% endfor %}
+            {% for file in platform.iter_files(".il") -%}
+                read_ilang {{file}}
+            {% endfor %}
+            read_ilang {{name}}.il
+            delete w:$verilog_initial_trigger
+            {{get_override("script_after_read")|default("# (script_after_read placeholder)")}}
+            synth_gowin {{get_override("synth_opts")|options}} -top {{name}} -json {{name}}.json
+            {{get_override("script_after_synth")|default("# (script_after_synth placeholder)")}}
+        """,
+    }
+
+    _apicula_command_templates = [
+        r"""
+        {{invoke_tool("yosys")}}
+            {{quiet("-q")}}
+            {{get_override("yosys_opts")|options}}
+            -l {{name}}.rpt
+            {{name}}.ys
+        """,
+        r"""
+        {{invoke_tool("nextpnr-gowin")}}
+            {{quiet("--quiet")}}
+            {{get_override("nextpnr_opts")|options}}
+            --log {{name}}.tim
+            --device {{platform._part}}
+            --family {{platform.device}}
+            --json {{name}}.json
+            --freq 27
```

Comment:
This argument shouldn't be there--I don't know what way is there to set constrains for the Gowin toolchain but it is what should be used.

@psentee Unfortunately replacing it with the frequency of the internal oscillator, while convenient for the development boards, isn't the right solution here eiher.

Reply:
I removed it and nextpnr-gowin ran fine without it. As I understand it, it allows to set the minimum frequency for any clock domain. Once one has more than one clock domain this parameter becomes useless anyhow. Manually checking the timing report is needed and advised in any case.

Reply:
Ideally, nextpnr-gowin would read clock constraints from a file. However, it looks like the Gowin port specifically has no way to do this: https://github.com/YosysHQ/nextpnr/issues/1191

Reply:
[constraints.md](https://github.com/YosysHQ/nextpnr/blob/master/docs/constraints.md) of nextpnr describes how to use the `--pre-pack` option for adding individual clock constraints.

Reply:
@whitequark you are right. There is no `--pre-pack` option for nextpnr-gowin. Oops.

Reply:
`--pre-pack` only exists if nextpnr was built with boost-python, which is not always the case, so we can't rely on it. YoWASP for example lacks it.

Reply:
Can we leave it like this for now?

Reply:
As in, without `--freq`? Yeah.